### DISCUSSION TITLE
Drawing polygons on separate layers

### DISCUSSION THREAD

**Discussion by domagojhack** [raphaelquast/EOmaps#89] [Q&A]
I am building an app that uses user input to draw polygons on EOmaps using a custom click callback.

<details><summary>Currently, I can achieve this with:</summary>

```
import geopandas as gpd
from eomaps import Maps
import matplotlib
from matplotlib import collections
import matplotlib.pyplot as plt

from shapely.geometry import Polygon
import geopandas as gpd


poly = []

m = Maps(layer="DEM")

data = m.read_file.GeoTIFF(
    "base_layers/dem/halasevo.tif")

m.set_shape.shade_raster()
m.set_data(**data)
m.plot_map(cmap="gist_earth")

a = [[]]


def map_click_callback(**kwargs):
    pos = kwargs["pos"]
    global m
    global poly
    global a

    poly.append(pos)

    m.set_shape.shade_raster()
    m.set_data(**data)
    m.plot_map(cmap="gist_earth")

    m.add_marker(
        xy=pos, radius=20, shape="ellipses",
        fc="none", ec="r", hatch="///", lw=2, n=100)

    print("col")
    for idx, i in enumerate(m.ax.collections[:]):
        print(idx, i)
        if type(i) == matplotlib.collections.PatchCollection:
            m.ax.collections.pop(idx)
            print("removed", idx)

    if len(poly) >= 3:
        shape = Polygon(poly)
        gdf = gpd.GeoDataFrame({"name": poly, "geometry": shape}, crs="4326")
        m.add_gdf(gdf, hatch="///", ec="b", alpha=.3)

    m.redraw()

m.cb.click.attach(map_click_callback)

m.show()
```
</details>

![Peek 2022-06-14 08-54](https://user-images.githubusercontent.com/2804094/173512781-b718be9f-4c03-4990-b200-973e9bb72719.gif)

Currently, I am drawing my polygons on the same layer by redrawing my layer with every click.
Ideally, I would like to have a separate layer for my polygon that is simultaneously visible with my base layer. I cannot wrap my head around the best solution for it. For now, my idea is to create a separate layer that contains the base map and my polygon.

Is there a right way to do this?

**Comment by raphaelquast** [comment]
Hey, yep that idea crossed my mind as well😄 
Yes there is in fact a better way to do this! 
It's a bit more complicated to implement, but the good news is I actually have a working example on this (it's been on my list of to-do's for a while but it has never been finished so far...)

Unfortunately I'm currently a bit short on time, but I'll try to upload the relevant pieces to a branch soon... would you be willing to finalize it together so that it can be added in an upcoming EOmaps release?  
(the code I have also supports exporting the polygons to shapefiles once they are drawn but the api is not very mature yet)

**Reply by domagojhack** [reply]
Of course, I am, but I am also a bit short on time lately so I will probably not be able to commit to it right away.

**Reply by raphaelquast** [reply]
Sure! There's no hurry at all!🙂 My goal here is just to avoid "re-inventing the wheel" all the time, so that tools and features that are realized by somebody go into the package at some time and help the community to evolve. 
I don't expect you to get it fully integrated in EOmaps, just to contribute the findings you make along the way so that It can be implemented properly sometime in the future!  I'll ping you here once I found the time to put it on github!

**Reply by raphaelquast** [reply]
@DomagojHack OK, I've uploaded my prove-of-concept for a proper EOmaps-drawer to the "draw" branch and opened a "work-in-progress" pull-request #90 .  I've also added some basic comments on how to use the feature at the moment in #90 .
In case you manage to build upon the work I did there so far I'd be grateful if you could add any improvements (e.g. as pull-requests to this branch) and I'll of course contribute and provide assistance along the way to make this feature a reality sometime in the future. 

At the moment really nothing of this should be seen as "fixed"... rather as a first try to see how well this could work (and I think it shows that it has the potential to be awesome 😄)
I'm open to discuss/change how the current interface is set-up and I'm happy on any constructive feedback you can provide!

**Reply by domagojhack** [reply]
I pulled the branch. Gonna start to experiment with drawer and keep you posted.

**Reply by raphaelquast** [reply]
Awesome! I just realized that your initial question was actually about how to put the polygons on separate layers... I've added a quick commit to implement that just now!


<details> <summary>Here's how it works at the moment:</summary>

```python
m = Maps(layer="base")
m.add_feature.preset.coastline()
m2 = m.new_layer(layer="second layer")
m2.add_feature.preset.ocean()

m.util.layer_selector()

# these draws will go onto the currently active layer
# (e.g. use the layer-selector widget to select the layer you want to draw on)
m.util.draw.circle()
m.util.draw.rectangle()

# this one will always end up on the "second layer"
# (you can draw on any layer but once you are finished it will be put on "second layer")
d1 = m.util.draw.new_poly(layer="second layer")
d1.polygon()
```

</details>

**Reply by domagojhack** [reply]
I just got back from fieldwork and continue doing my project that uses EOmaps.

**Reply by raphaelquast** [reply]
nice 🙂 I've merged all recent changes (e.g. v4.3) into the `draw` branch to keep it up to date
(there were no conflicting changes)
